### PR TITLE
[Bug][Java] defaultValues for date and date-time params

### PR DESCRIPTION
**Pull request by cachescrubber**
### Fixes

* https://github.com/OpenAPITools/openapi-generator/issues/11392

### PR checklist
 
- [x ] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [ x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [ x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [ x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (5.3.0), `6.0.x`
- [ x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

### CODE REVIEW

File: samples/client/petstore/spring-cloud-date-time/src/main/java/org/openapitools/api/DefaultApi.java [hunk 1]
```diff
@@ -82,8 +82,8 @@
         consumes = "application/x-www-form-urlencoded"
     )
     ResponseEntity<Void> updatePetWithForm(
-        @ApiParam(value = "A date path parameter", required = true) @PathVariable("date") @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate date,
-        @ApiParam(value = "Updated last vist timestamp") @RequestParam(value="visitDate", required=false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME) OffsetDateTime visitDate
+        @ApiParam(value = "A date path parameter", required = true, defaultValue = "Thu Jan 01 00:00:00 UTC 1970") @PathVariable("date") @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate date,
+        @ApiParam(value = "Updated last vist timestamp", defaultValue = "OffsetDateTime.parse("1971-12-19T11:39:57Z[UTC]", java.time.format.DateTimeFormatter.ISO_ZONED_DATE_TIME.withZone(java.time.ZoneId.systemDefault()))") @RequestParam(value="visitDate", required=false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME) OffsetDateTime visitDate
```
Comment (cachescrubber) [inline samples/client/petstore/spring-cloud-date-time/src/main/java/org/openapitools/api/DefaultApi.java:86]: This won't compile ...

File: samples/client/petstore/spring-cloud-date-time/src/main/java/org/openapitools/api/DefaultApi.java [hunk 2]
```diff
@@ -52,10 +52,10 @@
         value = "/thingy/{date}"
     )
     ResponseEntity<Void> get(
-        @ApiParam(value = "A date path parameter", required = true) @PathVariable("date") @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate date,
-        @NotNull @ApiParam(value = "A date-time query parameter", required = true) @Valid @RequestParam(value = "dateTime", required = true) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME) OffsetDateTime dateTime,
-        @ApiParam(value = "A date header parameter", required = true) @RequestHeader(value = "X-Order-Date", required = true) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate xOrderDate,
-        @ApiParam(value = "A date cookie parameter") @CookieValue("loginDate") @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate loginDate
+        @ApiParam(value = "A date path parameter", required = true, defaultValue = "Sat Jan 01 00:00:00 UTC 1972") @PathVariable("date") @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate date,
```
Comment (cachescrubber) [inline samples/client/petstore/spring-cloud-date-time/src/main/java/org/openapitools/api/DefaultApi.java:55]: DefaultValue is not formatted but java.util.Date#toString instead.

**Comment by cachescrubber** [comment]
@wing328 FYI. Who could help with the defaultValue date format issue?

**Comment by wing328** [comment]
Is `toDefaultValue` what you're looking for?

https://github.com/OpenAPITools/openapi-generator/blob/9e1972bb1d3a4f1e3f47c87613f4050feada0146/modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaCodegen.java#L896

I think you will need to override this method in the spring generator to address the issue above.

**Comment by cachescrubber** [comment]
Is stumbled up this line: https://github.com/OpenAPITools/openapi-generator/blob/9e1972bb1d3a4f1e3f47c87613f4050feada0146/modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractJavaCodegen.java#L984

What is the reasoning behind that?

**Comment by cachescrubber** [comment]
@wing328 The result of **toDefaultValue** is ignore for all parameter type other than queryParams. Why is that? Any hint?

**Comment by cachescrubber** [comment]
> What is the reasoning behind that?

Answer to self: probably used to generate Model classes, not for Api method Parameter.

**Comment by cachescrubber** [comment]
Relates to https://github.com/OpenAPITools/openapi-generator/pull/8577

**Comment by cachescrubber** [comment]
Slowly getting there. Need to find out why  **application/x-www-form-urlencoded** are treated like model parameter.

File: samples/client/petstore/spring-cloud-date-time/src/main/java/org/openapitools/api/DefaultApi.java (hunk 1 above)
Comment (cachescrubber) [reply samples/client/petstore/spring-cloud-date-time/src/main/java/org/openapitools/api/DefaultApi.java]: Fixed with 933662b8e4eb6c57f4502db3c105c464ce385771

**Comment by cachescrubber** [comment]
@wing328 I added some more tests to cover this issue. BTW, I think it might be okay to move

org.openapitools.codegen.languages.AbstractJavaCodegen#toDefaultParameterValue

to DefaultCodegen. WDYT?

**Comment by cachescrubber** [comment]
Surprisingly there are no samples affected by the last changes. Should I add date params with defaults to the more widely used examples?

**Comment by cachescrubber** [comment]
I did (local only). We would  stuff like the following (a groovy example)

```
    Date lastFeed = OffsetDateTime.parse("1973-12-19T11:39:57Z[UTC]", java.time.format.DateTimeFormatter.ISO_ZONED_DATE_TIME.withZone(java.time.ZoneId.systemDefault()))
    
    Date dateOfBirth = LocalDate.parse("2021-01-01")
```

so obviously the code in toDefaultValue() is not capable of supporting multiple Date libraries.

**Comment by cachescrubber** [comment]
Ich enhanced `toDefaultValue` to only return a default for **date** or **date-time** when **dateTimeLibrary** is **java8**. This way we can safely remove legacy date apis without introducing more bugs and broken samples in the meantime.

@wing328 I'm happy with the PR, please review.

**Comment by wing328** [comment]
cc @OpenAPITools/generator-core-team as the change covers default codegen: https://github.com/OpenAPITools/openapi-generator/pull/11536/files#diff-3a138675e8cb40943bcb00feb46edd6f6ee5c5306c29cc19f02c845a299c8658R6289

**Pull request merged**